### PR TITLE
[WFCORE-3608] Add core pool size attribute to IO subsystem

### PR DESCRIPTION
**Pull request by dmlloyd**
https://issues.jboss.org/browse/WFCORE-3608

### CODE REVIEW

**Comment by dmlloyd** [comment]
Supercedes #3107

**Comment by dmlloyd** [comment]
I think I got some extra changes in here by mistake.  Checking...

File: io/subsystem/src/main/java/org/wildfly/extension/io/IOExtension.java [hunk 1]
```diff
@@ -65,14 +65,15 @@ public void initializeParsers(ExtensionParsingContext context) {
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, Namespace.IO_1_0.getUriString(), IOSubsystemParser_1_0::new);
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, Namespace.IO_1_1.getUriString(), IOSubsystemParser_1_1::new);
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, Namespace.IO_2_0.getUriString(), IOSubsystemParser_2_0::new);
+        context.setSubsystemXmlMapping(SUBSYSTEM_NAME, Namespace.IO_3_0.getUriString(), new IOSubsystemParser_3_0());
```
Comment (dmlloyd) [inline io/subsystem/src/main/java/org/wildfly/extension/io/IOExtension.java]: Fixed @jmesnil's comment about directly instantiating the latest parser.

File: io/subsystem/src/main/java/org/wildfly/extension/io/IOSubsystemTransformers.java [hunk 2]
```diff
@@ -44,10 +47,19 @@ public String getSubsystemName() {
     public void registerTransformers(SubsystemTransformerRegistration registration) {
         ChainedTransformationDescriptionBuilder chainedBuilder = TransformationDescriptionBuilder.Factory.createChainedSubystemInstance(registration.getCurrentSubsystemVersion());
 
-        // Current 3.0.0 to 2.0.0, aka EAP 7.0.0
-        buildTransformers_2_0(chainedBuilder.createBuilder(registration.getCurrentSubsystemVersion(), VERSION_2_0));
+        // Current 3.0.0 to 2.0.0, aka EAP 7.1.0
```
Comment (dmlloyd) [inline io/subsystem/src/main/java/org/wildfly/extension/io/IOSubsystemTransformers.java]: To @jmesnil's comment on #3107 - I think I've untangled the knot here but I'm not 100% sure.  A transformers expert should check this out to be sure.

File: io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTransformerTestCase.java [hunk 3]
```diff
@@ -53,7 +52,7 @@ public IOSubsystemTransformerTestCase() {
 
     @Override
     protected String getSubsystemXml() throws IOException {
-        return readResource("io-2.0-transformer.xml");
+        return readResource("io-3.0-transformer.xml");
```
Comment (dmlloyd) [inline io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTransformerTestCase.java]: I kept the original name scheme of the XML resources.

File: io/subsystem/src/main/resources/org/wildfly/extension/io/LocalDescriptions.properties [hunk 4]
```diff
@@ -4,6 +4,8 @@ io.remove=Operation Removes subsystem
 io.worker=Defines workers
 io.worker.add=Adds new worker
 io.worker.remove=Removes worker
+io.worker.task-core-threads=Specify the starting number of threads for the worker task thread pool.\
+  If not set, it defaults to 2.
```
Comment (dmlloyd) [inline io/subsystem/src/main/resources/org/wildfly/extension/io/LocalDescriptions.properties]: Fixed in accordance with @rsvoboda's comment on https://github.com/wildfly/wildfly-core/pull/3107#discussion_r172507513.

**Comment by dmlloyd** [comment]
OK should be all fixed now, pending clean CI

**Comment by dmlloyd** [comment]
Looks like the transformer fixed were not correct. Will look into it tomorrow.

File: io/subsystem/src/main/java/org/wildfly/extension/io/IOSubsystemTransformers.java [hunk 5]
```diff
@@ -44,10 +47,19 @@ public String getSubsystemName() {
...
+        buildTransformers_3_0(chainedBuilder.createBuilder(VERSION_3_0, VERSION_2_0));
+        // 2.0.0 to 1.0.0, aka EAP 7.0.0
+        buildTransformers_2_0(chainedBuilder.createBuilder(VERSION_2_0, VERSION_1_0));
 
-        chainedBuilder.buildAndRegister(registration, new ModelVersion[]{VERSION_2_0});
+        chainedBuilder.buildAndRegister(registration, new ModelVersion[]{ VERSION_3_0, VERSION_2_0 });
```
Comment (jmesnil) [inline io/subsystem/src/main/java/org/wildfly/extension/io/IOSubsystemTransformers.java]: That does not seem correct, the version chain goes from the oldest to the newest ("current version excluded").
It should be:

```
chainedBuilder.buildAndRegister(registration, new ModelVersion[]{ VERSION_2_0, VERSION_3_0 });
```

File: io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTransformerTestCase.java [hunk 6]
```diff
@@ -68,11 +67,14 @@ public void testTransformerEAP710() throws Exception {
     }
 
     private KernelServices buildKernelServices(ModelTestControllerVersion controllerVersion, ModelVersion version, String... mavenResourceURLs) throws Exception {
-        return this.buildKernelServices(this.getSubsystemXml(), controllerVersion, version, mavenResourceURLs);
+        String xmlName = String.format("eap-%s-transformer.xml", controllerVersion.getRealVersionName());
```
Comment (jmesnil) [inline io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTransformerTestCase.java]: The test fails because there is no `eap-7.0.0-transformer.xml` or `eap-7.1.0-transformer.xml` files. 

I now understand why @ctomc renamed the `io-${subsystem-version}-transformer.xml` to `eap-${eap-version}-transformer.xml` in his PR :)

File: io/subsystem/src/main/java/org/wildfly/extension/io/IOSubsystemTransformers.java (hunk 5 above)
Comment (jmesnil) [reply io/subsystem/src/main/java/org/wildfly/extension/io/IOSubsystemTransformers.java]: I think the error in `IOSubsystemTransformerTestCase.testRejectingTransformersEAP_7_1_0` is caused by the wrong version chain order.

Comment (dmlloyd) [reply io/subsystem/src/main/java/org/wildfly/extension/io/IOSubsystemTransformers.java]: This didn't seem to fix the problem... I thought the transformers were meant to go from newer to older though (as per the comments)?  Maybe I need to swap the order in createBuilder() as well?

File: io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTransformerTestCase.java (hunk 6 above)
Comment (dmlloyd) [reply io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTransformerTestCase.java]: I fixed this using a switch to set up the correct IO version.

**Comment by dmlloyd** [comment]
So... what would you guys say if I were restore the transformer tests exactly as they were with the original PR and to @Ignore them, with a new issue to resolve the problem which can be worked by someone who has a might higher knowledge-of-transformers:list-of-backed-up-work ratio than I do?

**Comment by rsvoboda** [comment]
We shouldn't merge stuff which can have backward compatibility issues.

@jmesnil, any idea who could help with transformers stuff ?

**Comment by kabir** [comment]
@dmlloyd @rsvoboda I'll take a look at the transformers

File: io/subsystem/src/main/resources/org/wildfly/extension/io/LocalDescriptions.properties (hunk 4 above)
Comment (kabir) [reply io/subsystem/src/main/resources/org/wildfly/extension/io/LocalDescriptions.properties]: @rsvoboda @dmlloyd I disagree - the default-value in the r-r-d output already contains this information

**Comment by dmlloyd** [comment]
Holding until transformer tests are added.

**Comment by dmlloyd** [comment]
Should be all fixed now.

**Comment by dmlloyd** [comment]
@jmesnil is this OK now?

File: io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTestCase.java [hunk 7]
```diff
@@ -108,26 +104,5 @@ protected RunningMode getRunningMode() {
...
     }
 
-    protected static final OptionAttributeDefinition ENABLED_PROTOCOLS = OptionAttributeDefinition.builder("enabled-protocols", Options.SSL_ENABLED_PROTOCOLS)
-            .setRequired(false)
-            .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES)
-            .setAllowExpression(true)
-            .build();
-
-    @Test
-    public void testSequence() throws Exception {
```
Comment (jmesnil) [inline io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTestCase.java]: why did you remove that test?

**Comment by jmesnil** [comment]
@dmlloyd the PR looks good except for the removed test `testSequence`. It seems unrelated to the new attribute, why did you remove it?

File: io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTestCase.java (hunk 7 above)
Comment (dmlloyd) [reply io/tests/src/test/java/org/wildfly/extension/io/IOSubsystemTestCase.java]: I think this must have been removed as a part of the original PR.  I could try adding it back.

**Comment by dmlloyd** [comment]
@jmesnil OK all passed with the missing test restored.

**Comment by jmesnil** [comment]
@dmlloyd it seems you pushed the wrong commit to this PR. it now contains the commit to upgrade  wildfly common 1.3.2.Final
